query/src/sql/optimizer/s_expr.rs

```diff
@@ -27,6 +29,10 @@ pub struct SExpr {
     children: Vec<SExpr>,
 
     original_group: Option<IndexType>,
+
+    /// A bitmap to record applied rules on current SExpr, to prevent
+    /// redundant transformations.
+    pub(super) applied_rules: AppliedRules,
```

Comment:
@andylokandy I'd like to use a interior mutability pattern here, since it's a interior state. But can not find a proper way. How do you think of this?